File: NorthAx/Networking/APIError.swift
import Foundation

/// Typed representation of the backend's error envelope (§11):
/// `{ "error": { "code", "message", "status" } }`.
///
/// `code` is the machine-readable string (e.g. `METRICS_NOT_FOUND`) so the UI
/// can react to specific conditions rather than parsing messages.
struct APIError: Error, Equatable {
    var code: String
    var message: String
    var status: Int

    /// Network/transport failure with no HTTP response (offline, timeout).
    static let offline = APIError(code: "OFFLINE", message: "No connection.", status: 0)
    /// Response could not be decoded.
    static let decoding = APIError(code: "DECODING_ERROR", message: "Unexpected response.", status: 0)

    var isUnauthorized: Bool { status == 401 }
    var isNotFound: Bool { status == 404 }

    // Known codes worth branching on in the UI.
    var isTokenExpired: Bool { code == "AUTH_TOKEN_EXPIRED" }
    var isTokenRevoked: Bool { code == "AUTH_TOKEN_REVOKED" }
    var isMetricsMissing: Bool { code == "METRICS_NOT_FOUND" }
    var isIntervalsNotConnected: Bool { code == "INTERVALS_NOT_CONNECTED" }
    var isAIUnavailable: Bool { code == "AI_UNAVAILABLE" }

    var userMessage: String { message }
}

/// Wire shape of the error envelope for decoding.
struct APIErrorEnvelope: Decodable {
    struct Body: Decodable {
        let code: String
        let message: String
        let status: Int
    }
    let error: Body
}

File: android/app/src/main/java/app/northax/data/remote/ApiError.kt
package app.northax.data.remote

import kotlinx.serialization.Serializable

/**
 * Typed representation of the backend's error envelope:
 * `{ "error": { "code", "message", "status" } }`.
 *
 * `code` is the machine-readable string (e.g. `METRICS_NOT_FOUND`) so the UI
 * can react to specific conditions rather than parsing messages.
 */
class ApiError(
    val code: String,
    val message2: String,
    val status: Int,
) : Exception(message2) {

    val isUnauthorized: Boolean get() = status == 401
    val isNotFound: Boolean get() = status == 404

    // Known codes worth branching on in the UI.
    val isIntervalsNotConfigured: Boolean get() = code == "INTERVALS_NOT_CONFIGURED"

    val userMessage: String get() = message2

    companion object {
        /** Network/transport failure with no HTTP response (offline, timeout). */
        fun offline() = ApiError("OFFLINE", "No connection.", 0)

        /** Response could not be decoded. */
        fun decoding() = ApiError("DECODING_ERROR", "Unexpected response.", 0)
    }
}

/** Wire shape of the error envelope for decoding. */
@Serializable
data class ApiErrorEnvelope(val error: Body) {
    @Serializable
    data class Body(val code: String, val message: String, val status: Int)
}
